### PR TITLE
Add clippy to CI

### PR DESCRIPTION
**Pull request by kirushik**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -21,9 +21,14 @@ matrix:
...
   - rust: nightly
 
+
+before_script:
+  - rustup component add clippy
+
 script:
   - cargo build --all
   - cargo test --all
+  - ./clippy.sh
```
Comment (ordian) [inline .travis.yml]: We could use cargo config to make an alias: [example](https://github.com/rust-analyzer/rust-analyzer/blob/master/.cargo/config).

File: core/src/io.rs [hunk 2]
```diff
@@ -93,14 +93,16 @@ pub struct MetaIoHandler<T: Metadata, S: Middleware<T> = middleware::Noop> {
 
 impl<T: Metadata> Default for MetaIoHandler<T> {
 	fn default() -> Self {
-		MetaIoHandler::with_compatibility(Default::default())
+		Self::with_compatibility(Compatibility::default())
 	}
 }
 
+//We operate on complex parametrized types here, makes sense to let compiler do the heavy typing excercises
+#[allow(clippy::default_trait_access)]
```
Comment (ordian) [inline core/src/io.rs]: maybe allow it in the whole module with `!`? Also, IMHO [it's](https://rust-lang.github.io/rust-clippy/v0.0.212/index.html#default_trait_access) not bad.

File: macros/src/pubsub.rs [hunk 3]
```diff
@@ -14,118 +14,118 @@ pub use self::pubsub::SubscriptionId;
 /// New PUB-SUB subcriber.
 #[derive(Debug)]
 pub struct Subscriber<T, E = core::Error> {
-	subscriber: pubsub::Subscriber,
-	_data: PhantomData<(T, E)>,
+    subscriber: pubsub::Subscriber,
```
Comment (ordian) [inline macros/src/pubsub.rs]: please revert this indentation change

**Comment by kirushik** [comment]
I think we should only run Clippy on `stable`.
I'll fix this tomorrow

File: derive/tests/macros.rs [hunk 4]
```diff
@@ -2,6 +2,7 @@ use serde_json;
 use jsonrpc_derive::rpc;
 use jsonrpc_core::{IoHandler, Response};
 
+#[allow(clippy::empty_enum)]
```
Comment (ordian) [inline derive/tests/macros.rs]: Would changing it to `!` be breaking? I think it worth doing eventually.

File: server-utils/src/cors.rs [hunk 5]
```diff
@@ -103,6 +103,9 @@ impl ops::Deref for Origin {
 	}
 }
 
+// FIXME: There's a potential for allocation optimization here, by moving to `Value(Box<Origin>)`
```
Comment (tomusdrw) [inline server-utils/src/cors.rs]: IMHO that makes no sense at all. It's again making the public API less ergonomic to make the linter happy

File: http/src/handler.rs [hunk 6]
```diff
@@ -351,10 +355,10 @@ impl<M: Metadata, S: Middleware<M>> RpcHandler<M, S> {
 			// Validate the ContentType header
 			// to prevent Cross-Origin XHRs with text/plain
 			Method::POST if Self::is_json(request.headers().get("content-type")) => {
-				let uri = if self.rest_api != RestApi::Disabled { Some(request.uri().clone()) } else { None };
+				let uri = if self.rest_api == RestApi::Disabled { None } else { Some(request.uri().clone()) };
 				RpcHandlerState::ReadingBody {
 					metadata,
-					request: Default::default(),
+					request: Vec::<u8>::default(),
```
[the remark this answers is not in the document]
Comment (tomusdrw) [reply http/src/handler.rs]: Maybe just use `vec![]` if that makes clippy more happy, but IMHO using `Default` like that makes complete sense, cause there is less places where you have to name the type.

File: ws/src/server.rs [hunk 7]
```diff
@@ -38,6 +38,9 @@ impl Server {
 		&self.addr
 	}
 
+	// Yes it's indeed too many -- but this is how our API goes
+	// TODO: Let's refactor this into a ServerOptions struct for the next major release
```
Comment (tomusdrw) [inline ws/src/server.rs]: Configuration object will just mean more code with no gains (the whole thing could be made private). We have a builder pattern implemented anyway.

File: stdio/src/lib.rs [hunk 8]
```diff
@@ -62,11 +62,11 @@ impl ServerBuilder {
...
 /// Process a request asynchronously
 fn process(io: &Arc<IoHandler>, input: String) -> impl Future<Item = String, Error = ()> + Send {
-	io.handle_request(&input).map(move |result| match result {
-		Some(res) => res,
-		None => {
-			info!("JSON RPC request produced no response: {:?}", input);
-			String::from("")
-		}
+	io.handle_request(&input).map(move |result|
+	if let Some(res) = result {
```
Comment (tomusdrw) [inline stdio/src/lib.rs]: Incorrect indentation

File: server-utils/src/stream_codec.rs [hunk 9]
```diff
@@ -98,10 +98,8 @@ impl Decoder for StreamCodec {
 
 				if depth == 0 && idx != start_idx && idx - start_idx + 1 > whitespaces {
 					let bts = buf.split_to(idx + 1);
-					match String::from_utf8(bts.as_ref().to_vec()) {
-						Ok(val) => { return Ok(Some(val)) },
-						Err(_) => { return Ok(None); } // skip non-utf requests (TODO: log error?)
-					};
+					// skip non-utf requests (TODO: log error?)
+					return Ok(String::from_utf8(bts.as_ref().to_vec()).ok());
```
Comment (tomusdrw) [inline server-utils/src/stream_codec.rs]: To ba fair I find the previous version more readable :)

**Comment by kirushik** [comment]
@tomusdrw @ordian @ascjones I'm terribly sorry, I will only have time to go through all this feedback and fix it tomorrow :disappointed:

**Comment by tomusdrw** [comment]
@kirushik any updates? Should someone take over?

**Comment by ascjones** [comment]
Closing as stale, but still something we want to do. 

Anyone feel free to pick this up and open a new PR.

**Pull request closed without merging**